File: python/browser-session_a423c9.py
#!/usr/bin/env python3
# browser-session.ps1 — portiert nach python
# Quelle: powershell, Projects@abstractions:powershell/browser-session.ps1
# Erzeugt: 2026-08-08 durch ABSTRACTIONS_MANAGER.py

# browser-session.py — portiert von powershell
# Quelle: powershell, browser-session.ps1
# Erzeugt: 2026-08-07 durch ABSTRACTIONS_MANAGER.py

"""
.SYNOPSIS
Persistente Browser-Sitzung der Sandbox.

.DESCRIPTION
Zweck: Plattformen ohne (nutzbare) API — WaveSpeed-Konsole, Perplexity,
Canva, Stock-Portale — erfordern einen echten Web-Login. Diese Sitzung
speichert Cookies/LocalStorage DAUERHAFT in einem user-data-dir, akzeptiert
Cookie-Banner automatisch und bleibt über Skript-Läufe hinweg angemeldet.

Profil-Verzeichnis: <repo>/.browser-profile (gitignored — enthält Secrets).

Nutzung (immer unter Xvfb, damit echtes Chrome mit Codecs läuft):
  xvfb-run -a python3 browser-session.py open <URL>          # öffnen, Cookies akzeptieren, Screenshot
  xvfb-run -a python3 browser-session.py login <URL> [--user-field ..] [--pass-field ..] [--env-user X] [--env-pass Y]
  xvfb-run -a python3 browser-session.py shot <URL> [--out file.png] [--wait ms] [--full]
  xvfb-run -a python3 browser-session.py state                 # gespeicherte Cookies auflisten (Domains)

Die Sitzung wird NICHT geschlossen-und-verworfen: das Profil bleibt auf Platte.
"""

import argparse
import asyncio
import os
import re
import sys
import time
from pathlib import Path
from urllib.parse import urlparse

from playwright.async_api import async_playwright


def get_flag_value(args, name, default=None):
    """Hilfsfunktion für Parameterverarbeitung"""
    flag_name = f"--{name}"
    if flag_name in args.rest:
        idx = args.rest.index(flag_name)
        if idx + 1 < len(args.rest):
            return args.rest[idx + 1]
    return default


def has_flag(args, name):
    """Prüft ob ein Flag gesetzt ist"""
    return f"--{name}" in args.rest


def load_env(repo_path):
    """Umgebungsvariablen laden"""
    env_path = Path(repo_path) / ".env"
    env_vars = {}
    if env_path.exists():
        with open(env_path, 'r') as f:
            for line in f:
                match = re.match(r'^\s*([A-Z0-9_]+)\s*=\s*"?([^"]*)"?$', line.strip())
                if match:
                    env_vars[match.group(1)] = match.group(2)
    return env_vars


async def accept_cookies(page):
    """Cookie Consent akzeptieren"""
    labels = [
        "Accept all", "Accept All", "Alle akzeptieren", "Accept all cookies",
        "Alle Cookies akzeptieren", "I agree", "Ich stimme zu", "Zustimmen",
        "Allow all", "Akzeptieren", "Accept", "Got it", "Agree"
    ]
    
    for name in labels:
        try:
            btn = await page.get_by_role("button", name=name).first
            if await btn.is_visible(timeout=800):
                await btn.click(timeout=1500)
                return name
        except Exception:
            pass
    
    # Generische Consent-IDs
    selectors = ["#onetrust-accept-btn-handler", "[aria-label*='accept' i]", "button[title*='accept' i]"]
    for sel in selectors:
        try:
            el = await page.query_selector(sel)
            if el and await el.is_visible(timeout=500):
                await el.click(timeout=1500)
                return sel
        except Exception:
            pass
    return None


async def main():
    parser = argparse.ArgumentParser(description="Browser Session Manager")
    parser.add_argument("cmd", nargs="?", help="Befehl: open, shot, login, state")
    parser.add_argument("target", nargs="?", help="Ziel URL")
    parser.add_argument("rest", nargs="*", help="Zusätzliche Parameter")
    
    args = parser.parse_args()
    
    # Pfad-Konfiguration
    script_path = Path(__file__).resolve()
    repo = script_path.parent.parent
    profile_dir = os.environ.get("BROWSER_PROFILE_DIR", str(repo / ".browser-profile"))
    chrome_paths = ["/usr/bin/google-chrome-stable", "/usr/bin/google-chrome"]
    chrome_executable = next((p for p in chrome_paths if Path(p).exists()), None)
    
    # Profil-Verzeichnis erstellen
    Path(profile_dir).mkdir(parents=True, exist_ok=True)
    
    # Proxy-Konfiguration
    socks = get_flag_value(args, "socks")
    if socks:
        proxy_url = f"socks5://{socks}"
    else:
        proxy_url = os.environ.get("HTTPS_PROXY") or os.environ.get("https_proxy")
    
    # Browser-Kontext starten
    browser_args = [
        "--no-sandbox",
        "--autoplay-policy=no-user-gesture-required",
        "--disable-blink-features=AutomationControlled"
    ]
    
    if proxy_url:
        browser_args.append("--ssl-version-max=tls1.2")
    
    async with async_playwright() as p:
        browser = await p.chromium.launch_persistent_context(
            user_data_dir=profile_dir,
            headless=False,
            executable_path=chrome_executable,
            viewport={"width": 1440, "height": 900},
            accept_downloads=True,
            ignore_https_errors=has_flag(args, "insecure"),
            proxy={"server": proxy_url} if proxy_url else None,
            args=browser_args
        )
        
        page = browser.pages[0] if browser.pages else await browser.new_page()
        
        try:
            if args.cmd == "state":
                cookies = await browser.cookies()
                domains = sorted(list(set(c['domain'] for c in cookies)))
                print(f"Profil: {profile_dir}")
                print(f"{len(cookies)} Cookies über {len(domains)} Domains:")
                for domain in domains:
                    print(f"  {domain}")
            
            elif args.cmd in ["open", "shot"]:
                if not args.target:
                    raise ValueError("URL fehlt")
                
                await page.goto(args.target, wait_until="domcontentloaded", timeout=60000)
                wait_time = int(get_flag_value(args, "wait", "2500"))
                await asyncio.sleep(wait_time / 1000)
                accepted = await accept_cookies(page)
                if accepted:
                    print(f"Cookie-Consent bestätigt via: {accepted}")
                await asyncio.sleep(1)
                
                out_file = get_flag_value(args, "out", f"/tmp/browser-{int(time.time()*1000000)}.png")
                full_page = has_flag(args, "full")
                await page.screenshot(path=out_file, full_page=full_page)
                print(f"Screenshot: {out_file}")
                print(f"URL final: {page.url}")
            
            elif args.cmd == "login":
                if not args.target:
                    raise ValueError("URL fehlt")
                
                env_vars = load_env(repo)
                user = env_vars.get(get_flag_value(args, "env-user", ""), get_flag_value(args, "user", ""))
                password = env_vars.get(get_flag_value(args, "env-pass", ""), get_flag_value(args, "pass", ""))
                
                await page.goto(args.target, wait_until="domcontentloaded", timeout=60000)
                await asyncio.sleep(2.5)
                await accept_cookies(page)
                
                if user:
                    uf = get_flag_value(args, "user-field", "input[type=email], input[name=email], input[name=username], input[id*=email i]")
                    user_field = await page.query_selector(uf)
                    if user_field:
                        await user_field.fill(user, timeout=8000)
                
                if password:
                    pf = get_flag_value(args, "pass-field", "input[type=password]")
                    pass_field = await page.query_selector(pf)
                    if pass_field:
                        await pass_field.fill(password, timeout=8000)
                
                out_file = get_flag_value(args, "out", f"/tmp/login-{int(time.time()*1000000)}.png")
                await page.screenshot(path=out_file)
                user_set = "gesetzt" if user else "-"
                pass_set = "gesetzt" if password else "-"
                print(f"Login-Formular ausgefüllt (user={user_set}, pass={pass_set}). Screenshot: {out_file}")
                print("Absenden bewusst NICHT automatisch — nächster Schritt nach Sichtprüfung.")
            
            else:
                print("Befehle: open <URL> | shot <URL> | login <URL> | state")
        
        finally:
            await browser.close()  # Profil (Cookies) bleibt auf Platte erhalten


if __name__ == "__main__":
    asyncio.run(main())

File: javascript/browser-session_a423c9.js
#!/usr/bin/env node
// browser-session.ps1 — portiert nach javascript
// Quelle: powershell, Projects@abstractions:powershell/browser-session.ps1
// Erzeugt: 2026-08-08 durch ABSTRACTIONS_MANAGER.py

// browser-session.js — portiert nach JavaScript für Node 20
// Quelle: powershell, browser-session.ps1
// Erzeugt: 2026-08-07 durch ABSTRACTIONS_MANAGER.py

/**
 * SYNOPSIS
 * Persistente Browser-Sitzung der Sandbox.
 *
 * DESCRIPTION
 * Zweck: Plattformen ohne (nutzbare) API — WaveSpeed-Konsole, Perplexity,
 * Canva, Stock-Portale — erfordern einen echten Web-Login. Diese Sitzung
 * speichert Cookies/LocalStorage DAUERHAFT in einem user-data-dir, akzeptiert
 * Cookie-Banner automatisch und bleibt über Skript-Läufe hinweg angemeldet.
 *
 * Profil-Verzeichnis: <repo>/.browser-profile (gitignored — enthält Secrets).
 *
 * Nutzung (immer unter Xvfb, damit echtes Chrome mit Codecs läuft):
 *   xvfb-run -a node browser-session.js open <URL>          // öffnen, Cookies akzeptieren, Screenshot
 *   xvfb-run -a node browser-session.js login <URL> [--user-field ..] [--pass-field ..] [--env-user X] [--env-pass Y]
 *   xvfb-run -a node browser-session.js shot <URL> [--out file.png] [--wait ms] [--full]
 *   xvfb-run -a node browser-session.js state               // gespeicherte Cookies auflisten (Domains)
 *
 * Die Sitzung wird NICHT geschlossen-und-verworfen: das Profil bleibt auf Platte.
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import playwright from 'playwright';
import dotenv from 'dotenv';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Hilfsfunktionen für Parameterverarbeitung
function getFlagValue(args, name, defaultValue = null) {
    const index = args.indexOf(`--${name}`);
    if (index !== -1 && index + 1 < args.length) {
        return args[index + 1];
    }
    return defaultValue;
}

function hasFlag(args, name) {
    return args.includes(`--${name}`);
}

// Umgebungsvariablen laden
function loadEnv(repoPath) {
    const envPath = path.join(repoPath, '.env');
    const result = {};
    if (fs.existsSync(envPath)) {
        const content = fs.readFileSync(envPath, 'utf-8');
        const lines = content.split('\n');
        for (const line of lines) {
            const match = line.match(/^\s*([A-Z0-9_]+)\s*=\s*"?([^"]*)"?\s*$/);
            if (match) {
                result[match[1]] = match[2];
            }
        }
    }
    return result;
}

// Cookie Consent akzeptieren
async function acceptCookies(page) {
    const labels = [
        "Accept all", "Accept All", "Alle akzeptieren", "Accept all cookies",
        "Alle Cookies akzeptieren", "I agree", "Ich stimme zu", "Zustimmen",
        "Allow all", "Akzeptieren", "Accept", "Got it", "Agree"
    ];

    for (const name of labels) {
        try {
            const btn = await page.getByRole('button', { name }).first();
            if (await btn.isVisible({ timeout: 800 })) {
                await btn.click({ timeout: 1500 });
                return name;
            }
        } catch (err) {
            // Ignoriere Fehler
        }
    }

    // Generische Consent-IDs
    const selectors = [
        "#onetrust-accept-btn-handler",
        "[aria-label*='accept' i]",
        "button[title*='accept' i]"
    ];
    for (const sel of selectors) {
        try {
            const el = await page.locator(sel).first();
            if (await el.isVisible({ timeout: 500 })) {
                await el.click({ timeout: 1500 });
                return sel;
            }
        } catch (err) {
            // Ignoriere Fehler
        }
    }
    return null;
}

// Hauptskript
async function main() {
    const args = process.argv.slice(2);
    const cmd = args[0];
    const target = args[1];

    // Pfad-Konfiguration
    const scriptPath = __filename;
    const REPO = path.resolve(__dirname, '..', '..');
    const PROFILE_DIR = process.env.BROWSER_PROFILE_DIR || path.join(REPO, '.browser-profile');

    // Profil-Verzeichnis erstellen
    if (!fs.existsSync(PROFILE_DIR)) {
        fs.mkdirSync(PROFILE_DIR, { recursive: true });
    }

    // Chrome-Pfad finden
    const CHROME_PATHS = [
        '/usr/bin/google-chrome-stable',
        '/usr/bin/google-chrome'
    ];
    let CHROME = null;
    for (const p of CHROME_PATHS) {
        if (fs.existsSync(p)) {
            CHROME = p;
            break;
        }
    }

    // Proxy-Konfiguration
    let PROXY = getFlagValue(args, 'socks', null);
    if (PROXY) {
        PROXY = `socks5://${PROXY}`;
    } else {
        PROXY = process.env.HTTPS_PROXY || process.env.https_proxy || null;
    }

    // Browser-Kontext starten
    const browserArgs = [
        '--no-sandbox',
        '--autoplay-policy=no-user-gesture-required',
        '--disable-blink-features=AutomationControlled'
    ];

    if (PROXY) {
        browserArgs.push('--ssl-version-max=tls1.2');
    }

    const browser = await playwright.chromium.launchPersistentContext(PROFILE_DIR, {
        headless: false,
        executablePath: CHROME,
        viewport: { width: 1440, height: 900 },
        acceptDownloads: true,
        ignoreHTTPSErrors: hasFlag(args, 'insecure'),
        proxy: PROXY ? { server: PROXY, bypass: 'localhost,127.0.0.1,::1' } : undefined,
        args: browserArgs
    });

    const page = browser.pages().length > 0 ? browser.pages()[0] : await browser.newPage();

    try {
        if (cmd === 'state') {
            const cookies = await browser.cookies();
            const domains = [...new Set(cookies.map(c => c.domain))].sort();
            console.log(`Profil: ${PROFILE_DIR}`);
            console.log(`${cookies.length} Cookies über ${domains.length} Domains:`);
            domains.forEach(domain => console.log(`  ${domain}`));
        } else if (cmd === 'open' || cmd === 'shot') {
            if (!target) throw new Error('URL fehlt');
            await page.goto(target, { waitUntil: 'domcontentloaded', timeout: 60000 });
            const waitTime = parseInt(getFlagValue(args, 'wait', '2500'), 10);
            await new Promise(resolve => setTimeout(resolve, waitTime));
            const accepted = await acceptCookies(page);
            if (accepted) console.log(`Cookie-Consent bestätigt via: ${accepted}`);
            await new Promise(resolve => setTimeout(resolve, 1000));
            const out = getFlagValue(args, 'out', path.join('/tmp', `browser-${Date.now()}.png`));
            await page.screenshot({
                path: out,
                fullPage: hasFlag(args, 'full')
            });
            console.log(`Screenshot: ${out}`);
            console.log(`URL final: ${page.url()}`);
        } else if (cmd === 'login') {
            if (!target) throw new Error('URL fehlt');
            const envVars = loadEnv(REPO);
            const user = envVars[getFlagValue(args, 'env-user', '')] || getFlagValue(args, 'user', '');
            const pass = envVars[getFlagValue(args, 'env-pass', '')] || getFlagValue(args, 'pass', '');
            await page.goto(target, { waitUntil: 'domcontentloaded', timeout: 60000 });
            await new Promise(resolve => setTimeout(resolve, 2500));
            await acceptCookies(page);
            if (user) {
                const uf = getFlagValue(args, 'user-field', 'input[type=email], input[name=email], input[name=username], input[id*=email i]');
                const userField = await page.locator(uf).first();
                await userField.fill(user, { timeout: 8000 });
            }
            if (pass) {
                const pf = getFlagValue(args, 'pass-field', 'input[type=password]');
                const passField = await page.locator(pf).first();
                await passField.fill(pass, { timeout: 8000 });
            }
            const out = getFlagValue(args, 'out', path.join('/tmp', `login-${Date.now()}.png`));
            await page.screenshot({ path: out });
            const userSet = user.length > 0 ? 'gesetzt' : '-';
            const passSet = pass.length > 0 ? 'gesetzt' : '-';
            console.log(`Login-Formular ausgefüllt (user=${userSet}, pass=${passSet}). Screenshot: ${out}`);
            console.log('Absenden bewusst NICHT automatisch — nächster Schritt nach Sichtprüfung.');
        } else {
            console.log('Befehle: open <URL> | shot <URL> | login <URL> | state');
        }
    } finally {
        await browser.close(); // Profil (Cookies) bleibt auf Platte erhalten
    }
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});
